GH-3523 Cache mechanism for MemStatementIterator(s)

hmottestad opened the pull request:
GitHub issue resolved: #3523 <!-- add a Github issue number here, e.g #123. -->

Briefly describe the changes proposed in this PR:
 - functionality to measure when an MemStatementIterator should be cached
 - cache for MemStatementIterator(s)
 -  
<!-- short description of your change goes here -->

----
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md) for more details):

 - [ ] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [ ] I've added tests for the changes I made
 - [ ] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/main/.github/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [ ] I've [squashed](https://rdf4j.org/documentation/developer/squashing) my commits where necessary 
 - [ ] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change

hmottestad commented:
### After
```
Benchmark                            Mode  Cnt   Score   Error  Units
QueryBenchmark.complexQuery          avgt    5   6.617 ± 0.020  ms/op
QueryBenchmark.groupByQuery          avgt    5   2.026 ± 0.016  ms/op
QueryBenchmark.pathExpressionQuery1  avgt    5  25.105 ± 0.418  ms/op
QueryBenchmark.pathExpressionQuery2  avgt    5   1.443 ± 0.060  ms/op
```

hmottestad commented:
Also need to benchmark what cost this has for write operations.

JervenBolleman commented:
What happens if to queries run in parallel and access the same iterator?

hmottestad commented:
It's not the actual iterator that is cached, but an ArrayList with the results from the iterator. So if two threads both hit the cache then they will each get a new iterator backed by the list.

hmottestad commented:
### Before
```
Benchmark                                                     Mode  Cnt      Score     Error  Units
QueryBenchmark.complexQuery                                   avgt    5     47.596 ±   0.488  ms/op
QueryBenchmark.different_datasets_with_similar_distributions  avgt    5      3.232 ±   0.090  ms/op
QueryBenchmark.groupByQuery                                   avgt    5      1.983 ±   0.072  ms/op
QueryBenchmark.long_chain                                     avgt    5   1055.210 ±  32.642  ms/op
QueryBenchmark.lots_of_optional                               avgt    5    271.959 ±   8.758  ms/op
QueryBenchmark.minus                                          avgt    5  12103.994 ± 100.582  ms/op
QueryBenchmark.nested_optionals                               avgt    5    262.800 ±   5.527  ms/op
QueryBenchmark.pathExpressionQuery1                           avgt    5     23.529 ±   0.894  ms/op
QueryBenchmark.pathExpressionQuery2                           avgt    5      1.596 ±   0.102  ms/op
QueryBenchmark.query_distinct_predicates                      avgt    5    194.093 ±   9.729  ms/op
QueryBenchmark.simple_filter_not                              avgt    5     10.962 ±   0.539  ms/op
```

### After
```
Benchmark                                                     Mode  Cnt      Score     Error  Units
QueryBenchmark.complexQuery                                   avgt    5      6.758 ±   0.084  ms/op
QueryBenchmark.different_datasets_with_similar_distributions  avgt    5      3.324 ±   0.128  ms/op
QueryBenchmark.groupByQuery                                   avgt    5      2.056 ±   0.077  ms/op
QueryBenchmark.long_chain                                     avgt    5   1044.142 ±  31.316  ms/op
QueryBenchmark.lots_of_optional                               avgt    5    282.401 ±   7.528  ms/op
QueryBenchmark.minus                                          avgt    5  12203.275 ± 410.164  ms/op
QueryBenchmark.nested_optionals                               avgt    5    268.526 ±   6.198  ms/op
QueryBenchmark.pathExpressionQuery1                           avgt    5     23.477 ±   0.752  ms/op
QueryBenchmark.pathExpressionQuery2                           avgt    5      1.553 ±   0.123  ms/op
QueryBenchmark.query_distinct_predicates                      avgt    5    169.904 ±   1.297  ms/op
QueryBenchmark.simple_filter_not                              avgt    5     10.334 ±   0.221  ms/op
```

hmottestad commented:
So this doesn't really benefit the other queries much at the moment.

jeenbroekstra commented on core/sail/memory/src/main/java/org/eclipse/rdf4j/sail/memory/CachingMemStatementIteration.java:
```diff
@@ -0,0 +1,69 @@
+			throw ((X) e);
+		}
+		if (iterator.hasNext()) {
+			return iterator.next();
```
Ah hang on, I think I understand: this is not the original iterator you're returning from, but the iterator over the cached list of results? Perhaps we can make that a bit more explicit in the variable naming.

jeenbroekstra reviewed (commented):
Looks good. Some internal comments while I was trying to puzzle my way through what was going on - some may be more relevant than others.

hmottestad commented:
I'll go over the naming a bit. Maybe something like CacheAwareIteration would be more informative. Also need to check how much this impacts data insertion.

hmottestad commented:
@jeenbroekstra Could you take another look? I've moved around some of the code and renamed things to make it all a bit more obvious what the code is supposed to do.

jeenbroekstra reviewed (approved):
Thanks @hmottestad, looks good to me, and much clearer what's going on I think. 

I was going to suggest to annotate the `MemStatementIteratorCache` as `@InternalUseOnly` but I believe we may already have that annotation at the package level, so I guess that's fine.
